src/Riok.Mapperly/Descriptors/InlineExpressionRewriter.cs

```diff
@@ -66,6 +66,38 @@ public override SyntaxNode VisitTypeArgumentList(TypeArgumentListSyntax node)
         return base.VisitMemberAccessExpression(node);
     }
 
+    public override SyntaxNode? VisitCastExpression(CastExpressionSyntax node)
+    {
+        var result = base.VisitCastExpression(node);
+
+        if (result is CastExpressionSyntax typedResult && semanticModel.GetSymbolInfo(node.Type).Symbol is ITypeSymbol namedTypeSymbol)
+        {
+            var fullyQualifiedType = FullyQualifiedIdentifier(namedTypeSymbol);
+
+            return typedResult.WithType(fullyQualifiedType.WithTriviaFrom(typedResult.Type));
+        }
+
+        return result;
+    }
+
+    public override SyntaxNode? VisitBinaryExpression(BinaryExpressionSyntax node)
+    {
+        var result = base.VisitBinaryExpression(node);
+
+        if (
+            result is BinaryExpressionSyntax typedResult
+            && typedResult.Kind() is SyntaxKind.AsExpression
```

Comment:
Nitpick: `typedResult.Kind() is SyntaxKind.AsExpression` could be simplified to `typedResult.IsKind(SyntaxKind.AsExpression)`